### PR TITLE
[Feature](mlu-ops): add dcn_forward and dcn_backward_weight op

### PR DESCRIPTION
**Pull request by SangChengC**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

add dcn_forward and dcn_backward_weight op

## 2. Modification
 
        modified:   docs/user_guide/9_operators/index.rst
        modified:   docs/MLU-OPS-OpList.md
        modified:   mlu_op.h
        modified:   test/mlu_op_gtest/pb_gtest/mlu_op_test_proto
        new file:   kernels/dcn_forward/dcn_forward.cpp
        new file:   kernels/dcn_forward/dcn_common.cpp
        new file:   kernels/dcn_forward/dcn_backward_weight.cpp
        new file:   test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp
        new file:   test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.h
        new file:   test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/dcn_backward_weight.cpp
        new file:   test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/dcn_backward_weight.h
        new file:   test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/test_case/case_hi_16.prototxt
        new file:   test/mlu_op_gtest/pb_gtest/src/zoo/dcn_forward/dcn_forward.cpp
        new file:   test/mlu_op_gtest/pb_gtest/src/zoo/dcn_forward/dcn_forward.h
        new file:   test/mlu_op_gtest/pb_gtest/src/zoo/dcn_forward/test_case/case_hi_16.prototxt

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [x] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [x] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590
- Job types
  - [ ] BLOCK
  - [x] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test (e.g. float32/int8)
- [x] Multi-dimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [x] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example: dcn_forward
#
# ----------- case0 -----------
[MLU Hardware Time      ]: 187 (us)
[MLU Interface Time     ]: 135.677 (us)
[MLU IO Efficiency      ]: 0.00457182
[MLU Compute Efficiency ]: 0.0293672
[MLU Workspace Size     ]: 5.52968e+06 (Bytes)
[MLU TheoryOps          ]: 1.26528e+07 (Ops)
[MLU TheoryIOs          ]: 1.7509e+06 (Bytes)
[MLU ComputeForce       ]: 2.304e+12 (op/s)
[MLU IoBandWidth        ]: 2048 (GB/s)
[GPU Hardware Time      ]: -1 (us)
[GPU IO Efficiency      ]: -1
[GPU Compute Efficiency ]: -1
[GPU Workspace Size     ]: -1 (Bytes)
[Diffs]:
[output]
DIFF1: 3.491183e-05
DIFF2: 3.435694e-05
[^      OK ] ../../test/mlu_op_gtest/pb_gtest/src/zoo/dcn_forward/test_case/case_hi_16.prototxt
[       OK ] dcn_forward/TestSuite.mluOp/0 (128 ms)
[----------] 1 test from dcn_forward/TestSuite (128 ms total)
# 
# for example: dcn_backward_weight
# ----------- case0 -----------
[MLU Hardware Time      ]: 229 (us)
[MLU Interface Time     ]: 32051.5 (us)
[MLU IO Efficiency      ]: 0.00373332
[MLU Compute Efficiency ]: 0.0239811
[MLU Workspace Size     ]: 5.54402e+06 (Bytes)
[MLU TheoryOps          ]: 1.26528e+07 (Ops)
[MLU TheoryIOs          ]: 1.7509e+06 (Bytes)
[MLU ComputeForce       ]: 2.304e+12 (op/s)
[MLU IoBandWidth        ]: 2048 (GB/s)
[GPU Hardware Time      ]: -1 (us)
[GPU IO Efficiency      ]: -1
[GPU Compute Efficiency ]: -1
[GPU Workspace Size     ]: -1 (Bytes)
[Diffs]:
[grad_weight]
DIFF1: 4.231463e-05
DIFF2: 4.225033e-05
[grad_bias]
DIFF1: 2.666710e-07
DIFF2: 3.025384e-07
[^      OK ] ../../test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/test_case/case_hi_16.prototxt
[       OK ] dcn_backward_weight/TestSuite.mluOp/0 (173 ms)
[----------] 1 test from dcn_backward_weight/TestSuite (173 ms total)
# ...
```

Platform：MLU590
```bash
# for example: dcn_forward
#
# ----------- case0 -----------
[MLU Hardware Time      ]: 187 (us)
[MLU Interface Time     ]: 135.677 (us)
[MLU IO Efficiency      ]: 0.00457182
[MLU Compute Efficiency ]: 0.0293672
[MLU Workspace Size     ]: 5.52968e+06 (Bytes)
[MLU TheoryOps          ]: 1.26528e+07 (Ops)
[MLU TheoryIOs          ]: 1.7509e+06 (Bytes)
[MLU ComputeForce       ]: 2.304e+12 (op/s)
[MLU IoBandWidth        ]: 2048 (GB/s)
[GPU Hardware Time      ]: -1 (us)
[GPU IO Efficiency      ]: -1
[GPU Compute Efficiency ]: -1
[GPU Workspace Size     ]: -1 (Bytes)
[Diffs]:
[output]
DIFF1: 3.491183e-05
DIFF2: 3.435694e-05
[^      OK ] ../../test/mlu_op_gtest/pb_gtest/src/zoo/dcn_forward/test_case/case_hi_16.prototxt
[       OK ] dcn_forward/TestSuite.mluOp/0 (128 ms)
[----------] 1 test from dcn_forward/TestSuite (128 ms total)
# 
# for example: dcn_backward_weight
# ----------- case0 -----------
[MLU Hardware Time      ]: 229 (us)
[MLU Interface Time     ]: 32051.5 (us)
[MLU IO Efficiency      ]: 0.00373332
[MLU Compute Efficiency ]: 0.0239811
[MLU Workspace Size     ]: 5.54402e+06 (Bytes)
[MLU TheoryOps          ]: 1.26528e+07 (Ops)
[MLU TheoryIOs          ]: 1.7509e+06 (Bytes)
[MLU ComputeForce       ]: 2.304e+12 (op/s)
[MLU IoBandWidth        ]: 2048 (GB/s)
[GPU Hardware Time      ]: -1 (us)
[GPU IO Efficiency      ]: -1
[GPU Compute Efficiency ]: -1
[GPU Workspace Size     ]: -1 (Bytes)
[Diffs]:
[grad_weight]
DIFF1: 4.231463e-05
DIFF2: 4.225033e-05
[grad_bias]
DIFF1: 2.666710e-07
DIFF2: 3.025384e-07
[^      OK ] ../../test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/test_case/case_hi_16.prototxt
[       OK ] dcn_backward_weight/TestSuite.mluOp/0 (173 ms)
[----------] 1 test from dcn_backward_weight/TestSuite (173 ms total)
# ...
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: mlu_op.h [hunk 1]
```diff
@@ -16784,6 +16784,463 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+
+ * @par API Dependency
+ * - Before calling this function to implement the backward filter of deformable convolution,
+ *   you need to prepare all the parameters passed to this function. See each parameter
+ *   description for details.
+ *
+ * @par Performance Optimization
+ * - For best practices, to have better performance, set the im2col_step to the batch size.
+ *
+ * @note
```
Comment (mahxn0) [inline mlu_op.h:17212]: 注释格式不对

**Review by mahxn0 — APPROVED** [review]

File: docs/user_guide/9_operators/index.rst [hunk 2]
```diff
@@ -1014,3 +1014,11 @@ mluOpConcat
 mluOpDCNBackwardData
 ---------------------------------
 该算子用于求取可变形卷积算子关于input、offset、mask的反向梯度。
+
+mluOpDCNForward
+-----------------------------
+可变形卷积。通过额外的offset和mask来增强滤波器对空间的几何表达能力,并且该卷积可以任意替代之前卷及网络里面的任意常规卷积层。
```
Comment (PetrelYy) [inline docs/user_guide/9_operators/index.rst:1020]: 添加类似 .. _dcn_backward_data: 的字段

File: kernels/dcn_common/dcn_common.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
+/*************************************************************************
```
Comment (DanieeelLiu) [inline kernels/dcn_common/dcn_common.cpp:1]: 建议挪到kernel/utils/下面，作为工具函数引用，而不是算子

File: mlu_op.h [hunk 4]
```diff
@@ -16784,6 +16784,463 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ * - The allocated extra workspace should be passed to the ::mluOpDCNForward function to perform
+ *   the deformable convolution forward operation.
+ *
+ * @par Note
+ * - None.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
```
Comment (PetrelYy) [inline mlu_op.h:16843]: 参考文档规范中格式，补充 reference等字段，文档部分私聊 @ suhaoming review

File: test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
+/*************************************************************************
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (PetrelYy) [inline test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp:10]: 统一修改，规范中有模板，修改Copyright 日期

File: mlu_op.h [hunk 6]
```diff
@@ -16768,7 +16768,7 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
  * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_EXECUTION_FAILED
  *
```
Comment (DanieeelLiu) [inline mlu_op.h:16769]: 文件权限 从100644 变  100755了，确认下是否需要更改

File: test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+#include "core/tensor.h"
+#include "kernels/kernel.h"
+#include "kernels/debug.h"
+
+#define TRANSPOSE_MAX_DIM 8
+
+
+
+mluOpStatus_t mluOpTransposeCpu(const mluOpTransposeDescriptor_t desc,
+                              const mluOpTensorDescriptor_t x_desc,
```
Comment (PetrelYy) [inline test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.h:25]: 使用根目录 tools/format 刷下格式

File: kernels/dcn_common/dcn_common.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#ifndef KERNELS_DCN_DCN_H_
+#define KERNELS_DCN_DCN_H_
```
Comment (Unireverse) [inline kernels/dcn_common/dcn_common.h:24]: 这个不规范，需要按照文件路径写宏，这里就要写成 KERNELS_DCN_COMMON_DCN_COMMON_H

File: kernels/dcn_common/dcn_common.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+#include <limits.h>
+#include <math.h>
+#include <vector>
+
+#include "kernels/utils/cnnl_helper.h"
+
+#define DCN_API "mluOpDcn"
```
Comment (Unireverse) [inline kernels/dcn_common/dcn_common.h:31]:
```suggestion
#define DCN_API "mluOpDCN"
```

File: test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+ *************************************************************************/
+
+#include "transpose_cpu.h"
+#include <vector>
+
+#include "core/tensor.h"
+#include "mlu_op.h"
+// #include "core/context.h"
```
Comment (PetrelYy) [inline test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp:18]: dcn cpu 用到这个文件了是吧？

File: kernels/dcn_forward/dcn_forward.cpp [hunk 11]
```diff
@@ -0,0 +1,104 @@
+/*************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (Unireverse) [inline kernels/dcn_forward/dcn_forward.cpp:2]: 新算子入库，协议中的时间要写入库的年份

File: test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+// #include "core/runtime/device.h"
+// #include "core/type.h"
+// #include "core/tool.h"
+struct cnnlTransposeStruct {
+  int64_t dim;
+  std::vector<int> permute;
+};
```
Comment (Unireverse) [inline test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp:25]: 为啥定义这个

File: test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+#include "transpose_cpu.h"
+#include <vector>
+
+#include "core/tensor.h"
+#include "mlu_op.h"
+// #include "core/context.h"
+// #include "core/runtime/device.h"
+// #include "core/type.h"
+// #include "core/tool.h"
+struct cnnlTransposeStruct {
```
[the remark this answers is not in the document]
Comment (SangChengC) [reply test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp:28]: 在mlu_op.h里是typedef struct cnnlTransposeStruct *mluOpTransposeDescriptor_t;

File: kernels/dcn_common/dcn_common.cpp (hunk 3 above)
Comment (SangChengC) [reply kernels/dcn_common/dcn_common.cpp:1]: 挪到前向里老是报重复引用的错误, 我看cnnl里也是加了个dcn_common,所以写成这样了

File: kernel_depends.toml [hunk 14]
```diff
@@ -37,6 +37,9 @@ masked_col2im_forward = ["fill", "transpose"]
 masked_im2col_forward = ["fill", "transpose"]
 mutual_information_backward = ["fill"]
 batch_matmul_bcast = ["matmul"]
+dcn_forward = ["dcn_common"]
```
Comment (mahxn0) [inline kernel_depends.toml:40]: 后面修改后这里删掉吧  dcn_common不是算子

File: kernels/dcn_common/dcn_common.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+mluOpStatus_t MLUOP_WIN_API
+mluOpDestroyDCNDescriptor(mluOpDCNDescriptor_t dcn_desc);
+
+mluOpStatus_t MLUOP_WIN_API mluOpSetDCNDescriptor(
+    mluOpDCNDescriptor_t dcn_desc, int dimNb, const int pad[],
+    const int stride[], const int dilation[], int deformable_group,
+    int conv_group, int im2col_step, const mluOpDataType_t compute_type);
+
+#endif  // KERNELS_DCN_COMMON_DCN_COMMON_H
```
Comment (mahxn0) [inline kernels/dcn_common/dcn_common.h:45]: 已经在mlu_op.h中申明  此处不用再额外申明，将dcn_common文件夹删掉，实现放在kernel/utils/dcn_common.h中就行

File: test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (mahxn0) [reply test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp:28]: 这里是仅在当前文件有效的结构体，建议换个命名方式。mluOpTransposeStruct似乎更合适

File: test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+#include "transpose_cpu.h"
+#include <vector>
+#include "core/tensor.h"
+
+struct cnnlTransposeStruct {
+  int64_t dim;
+  std::vector<int> permute;
+};
+// #include "kernels/transpose/transpose.h"
```
Comment (mahxn0) [inline test/mlu_op_gtest/pb_gtest/src/internal_kernel/transpose_cpu/transpose_cpu.cpp:32]: 不需要的注释删掉

File: test/mlu_op_gtest/pb_gtest/src/zoo/dcn_forward/test_case/case_hi_16.prototxt [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+  id: "input"
+  shape: {
+    dims: 1
+    dims: 16
+    dims: 16
+    dims: 300
+  }
+  layout: LAYOUT_NHWC
+  dtype: DTYPE_FLOAT
+  onchip_dtype: DTYPE_INT16
```
Comment (mahxn0) [inline test/mlu_op_gtest/pb_gtest/src/zoo/dcn_forward/test_case/case_hi_16.prototxt:13]: 同上

File: test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/test_case/case_hi_16.prototxt [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+  id: "grad_output"
+  shape: {
+    dims: 1
+    dims: 16
+    dims: 16
+    dims: 300
+  }
+  layout: LAYOUT_NHWC
+  dtype: DTYPE_FLOAT
+  onchip_dtype: DTYPE_INT16
```
Comment (mahxn0) [inline test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/test_case/case_hi_16.prototxt:66]: 同上

File: test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/test_case/case_hi_16.prototxt [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+  id: "input"
+  shape: {
+    dims: 1
+    dims: 16
+    dims: 16
+    dims: 300
+  }
+  layout: LAYOUT_NHWC
+  dtype: DTYPE_FLOAT
+  onchip_dtype: DTYPE_INT16
```
Comment (mahxn0) [inline test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_weight/test_case/case_hi_16.prototxt:13]: onchip_dtype删掉，后续不再支持片上量化。

File: kernels/dcn_forward/dcn_forward.cpp [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include "kernels/dcn_common/dcn_common.h"
+
+#define DCNFORWARD_API "mluOpDcnForward"
```
Comment (mahxn0) [inline kernels/dcn_forward/dcn_forward.cpp:25]: mluOpDCNForward 下同

**Review by mahxn0 — APPROVED** [review]

File: mlu_op.h [hunk 21]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ *   Input. Pointer to the MLU memory that stores the input tensor.
+ * @param[in] offset_desc
+ *   Input. The descriptor of the offset tensor to be applied for each position in the convolution kernel.
+    The shape of offset should be (batch, out_height, out_width, 2  deformable_group *
+    filter_height  filter_width). For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] offset
+ *   Input. Pointer to the MLU memory that stores the offset tensor.
+ * @param[in] mask_desc
+ *   Input. The descriptor of the scaling factor to be applied for each position in the convolution
+ *   kernel. The shape of mask should be (batch, out_height, out_width,
```
Comment (AquamarineSu) [inline mlu_op.h:16890]:
```suggestion
 *   kernel. The shape of the mask must be (batch, out_height, out_width,
```

File: mlu_op.h [hunk 22]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+    deformable_group  filter_height * filter_width). Set this parameter to NULL when
+ *   mask is not requested. For detailed information, see ::mluOpTensorDescriptor_t.
```
Comment (AquamarineSu) [inline mlu_op.h:16892]:
```suggestion
 *  the mask is not requested. For detailed information, see ::mluOpTensorDescriptor_t.
```

File: mlu_op.h [hunk 23]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ *   requested. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] bias
+ *   Input. Pointer to the MLU memory that stores the bias tensor. Set this parameter to NULL when bias is not
+ *   requested.
+ * @param[in] workspace
+ *   Input. Pointer to the MLU memory that is used as an extra workspace for the
+ *   deformable convolution operation. For more information about workspace, see
+ *   "Cambricon MLUOP User Guide".
+ * @param[in] workspace_size
+ *   Input. The size of the extra workspace in bytes that needs to be used in the deformable
```
Comment (AquamarineSu) [inline mlu_op.h:16912]:
```suggestion
 *   Input. The size of the extra workspace in bytes needed for the deformable
```

File: mlu_op.h [hunk 24]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ * @param[out] output
+ *   Output. Pointer to the MLU memory that stores the output tensor.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM,
+ *   ::MLUOP_STATUS_NOT_SUPPORTED, ::MLUOP_STATUS_NUMERICAL_OVERFLOW
+ * @par Formula
+ * - See "Deformable Convolution Operator" section in "Cambricon MLUOP User Guide" for details.
+ *
+ * @par Data Type
+ * - Offchip data type of \p input, \p offset, \p mask, \p filter, \p bias, and \p output must be the same.
```
Comment (AquamarineSu) [inline mlu_op.h:16927]:
```suggestion
 * - The off-chip data type of \p input, \p offset, \p mask, \p filter, \p bias, and \p output must be the same.
```

File: mlu_op.h [hunk 25]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ * - The supported offchip data types of the input tensor and output tensor are as follows:
```
Comment (AquamarineSu) [inline mlu_op.h:16928]:
```suggestion
 * - The supported off-chip data types of the input tensor and output tensor are as follows:
```

File: mlu_op.h [hunk 26]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ *   - input, offset, mask, filter, bias, output: half, float.
+ * - This function supports any combinations of the following onchip data types for input tensor
```
Comment (AquamarineSu) [inline mlu_op.h:16930]:
```suggestion
 * - This function supports any combinations of the following on-chip data types for input tensor
```

File: mlu_op.h [hunk 27]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ *   \p input and \p filter on MLU200 series and CE3226.
+ *   - \p input onchip data type: int16, int31.
+ *   - \p filter onchip data type: int16, int31.
+ * - \p input offchip data type can be combined with any supported onchip data types.
+ * - \p filter offchip data type can be combined with any supported onchip data types.
+ * - This function also supports float-point computation on MLU300 series or above.
```
Comment (AquamarineSu) [inline mlu_op.h:16936]:
```suggestion
 * - This function also supports floating-point computation on MLU300 series or above.
```

File: mlu_op.h [hunk 28]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+ *   To perform float-point computation, the onchip data type of \p input and \p filter
```
Comment (AquamarineSu) [inline mlu_op.h:16937]:
```suggestion
 *   To perform floating-point computation, the onchip data type of \p input and \p filter
```

File: mlu_op.h [hunk 29]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+      \p channel should be equal to deformable_group  filter_height * filter_width.
+    - The value of (im2col_step  out_height  out_filter  filter_h  filter_w  input_channel)
+ *     should be less than or equal to the INT_MAX defined in limits.h.
+ * @par API Dependency
+ * - Before calling this function to implement deformable convolution, you need to prepare
+ *   all the parameters passed to this function. See each parameter description
+ *   for details.
+ *
+ * @par Performance Optimization
+ * - For best practices, to have a better performance, set the im2col_step equal to the batch
```
Comment (AquamarineSu) [inline mlu_op.h:16965]:
```suggestion
 * - To achieve better performance, set the im2col_step equal to the batch
```

File: mlu_op.h [hunk 30]
```diff
@@ -16784,6 +16784,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
...
+                const mluOpTensorDescriptor_t bias_desc,
+                const void *bias,
+                void *workspace,
+                size_t workspace_size,
+                const mluOpTensorDescriptor_t output_desc,
+                void *output);
+
+// Group:DCN
+/*!
+ * @brief Returns in \p workspace_size the size of the MLU memory that is used as an extra
```
Comment (AquamarineSu) [inline mlu_op.h:17030]: 同上

File: mlu_op.h [hunk 31]
```diff
@@ -12070,7 +12070,7 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
  * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_EXECUTION_FAILED
  *
  * @par Note
- * - Call this function after calling the ::mluOpDCNBackwardData,
+ * - Call this function after calling the ::mluOpDCNBackwardData,mluOpDCNForward,
```
Comment (AquamarineSu) [inline mlu_op.h:12073]:
```suggestion
 * - Call this function after calling the ::mluOpDCNBackwardData, ::mluOpDCNForward,
```

**Review by AquamarineSu — APPROVED** [review]

**Review by mahxn0 — APPROVED** [review]

**Review by Unireverse — APPROVED** [review]

File: mlu_op.h [hunk 32]
```diff
@@ -12086,6 +12086,469 @@ mluOpSetDCNDescriptor(mluOpDCNDescriptor_t dcn_desc,
  */
 mluOpStatus_t MLUOP_WIN_API
 mluOpDestroyDCNDescriptor(mluOpDCNDescriptor_t dcn_desc);
+// Group:DCN
```
Comment (duzekunKTH) [inline mlu_op.h:12089]: 接口注释上面增加一行空行

**Pull request merged**